### ISSUE TITLE
FXIOS-1077 ⁃ Two new Tabs when using Apple‘s Magic Keyboard

### ISSUE THREAD

**Issue by m-him** [mozilla-mobile/firefox-ios#7519]
## Steps to reproduce
open firefox, open new tab by using shortcut Strg + T.
closing Tab (Strg + W)

### Expected behavior
opens a new tab when using the shortcut Strg + T.
duplicates a tab, when closing a tab with shortcur Strg + W.

### Actual behavior
opens two new tabs when using the shortcut Strg + T

### Device & build information
* Device: ipad pro (only by using the magic keyboard)
* Build version: 29
### Notes
Attachments:

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-2583)

**Comment by shrop** [comment]
Also seeing the following issue with an iPad Pro and the Magic Keyboard when using Command+R to refresh a page.

What happens: A new tab opens with the refreshed content of the previous tab

What's expected: The current tab refreshes page content. No new tabs opened.

**Comment by adamcstephens** [comment]
I'm having this problem as well. iPad Pro 10.5" with Magic Keyboard. Refreshing, closing, and sometimes even entering a URL can all cause this to happen. I've attached a screen recording.

I hit this bug frequently since I primarily use my tablet with the keyboard. It can be quite jarring and frustrating.

https://user-images.githubusercontent.com/2071575/106546642-ab53aa80-64d9-11eb-822a-398cdd9b5c49.MP4

**Comment by phkeese** [comment]
Same issue here with a Logitech Combo Touch on an 8th gen iPad. When I press CMD+T, sometimes two new tabs get created.
Some times one tab is created, but as soon as I type in an address, another tab with that address opens and the current tab becomes unresponsive.

**Comment by naavis** [comment]
I encounter the same issue all the time with an iPad Air and Apple Smart Keyboard. There is also a thread open about this on the support forum: https://support.mozilla.org/en-US/questions/1331571

**Comment by adamcstephens** [comment]
This issue is still present after the tab changes in the recent version. Personally, I stopped using Firefox on my iPad months ago because of this bug.

**Comment by danigol** [comment]
Same, I only open Firefox anymore to see if they fixed this bug. Months now. I've just quit using Firefox on iPad. One of the core reasons I started using Firefox everywhere was the fact that it _is_ everywhere. This has been making me question using it on other platforms as well.

**Comment by skabber** [comment]
This may seem like a minor bug, but it is so infuriating that I am sad to have moved on to Chrome.

**Comment by SimonBasca** [comment]
Comment from #8707

Note to Devs. While I don't have a real hardware keyboard on a real iPad to verify my findings, it is possible to replicate this in the Simulator. At which point the Logic error comes to light:

Assuming that in real life the way the key presses happen is:

    Command Key goes down
    Shortly after the 'T' key goes down
    Shortly after the 'T' key is released
    Shortly after the Command key is released

At stage 1, and this might be a surprise to you, this code in BrowserViewController triggers:

     override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
         super.pressesBegan(presses, with: event)
         if #available(iOS 13.4, *) {
             guard let key = presses.first?.key, (key.keyCode == .keyboardLeftGUI || key.keyCode == .keyboardRightGUI) else { return } //GUI buttons = CMD buttons on ipad/mac
             self.isCmdClickForNewTab = true
         }
    }

Then at Stage 2, this then fires from TabTrayController+KeyCommands:

    @objc func didOpenNewTabKeyCommand() {
        TelemetryWrapper.recordEvent(category: .action, method: .press, object: .keyCommand, extras: ["action": "new-tab"])
        openNewTab()
    }

and a new tab is created.

When the new tab is created, BrowserViewController.webview:decidePolicy... fires and hits this bit of code:

        if isCmdClickForNewTab {
            guard let url = webView.url, let isPrivate = self.tabManager.selectedTab?.isPrivate else { return }
            homePanelDidRequestToOpenInNewTab(url, isPrivate: isPrivate)
            self.isCmdClickForNewTab = false
            decisionHandler(.cancel)
        }

And because isCmdClickForNewTab was set in Stage 1 another tab is created.

Hence why we end up with 2 tabs instead of 1.

In the Simulator I have also seen cases where multiple tabs will be opened until you get past stage 4 above. But because Cmd-key handling in Simulator is a bit flakey I really don't know if this is a real world possibility.

Note that you will also get 2 tabs if you hold down the Command Key while pressing the "Create New Tab" '+' button in the top-right of the screen.

So to summarise - I don't think pressesBegin/End do what you expect. Looking at how all this comes together I think you think that those will only be called when the user physically touches the screen. But as we have seen, touching keys on the hardware keyboard also triggers them.

**Comment by sabberworm** [comment]
I have this too and it’s annoying as hell. I have the keybaord/trackpad connected all the time and it happens also when opening links and closing tabs, even when not using the keyboard (but only when it’s attached). It happens on a fresh install. Makes Firefox almost unusable for me.

**Comment by adamcstephens** [comment]
> **Update**: It looks like this doesn’t happen when I close tabs with the close icon instead of the cmd+W shortcut. I will test for a few days and report back.

This is my experience as well. The issue only occurs with keyboard shortcuts, not the touchscreen interface.

**Comment by SimonBasca** [comment]
Same issue is happening when opening a new private tab, when pressing "cmd + ]" to navigate forward and when pressing "cmd+R" to reload a page.

**Comment by lokan** [comment]
So this is a bug since Oct 2020 and there is still no fix for it? Looks pretty easy to patch though, no? I want to use Firefox, but if I have to do three actions instead of one, I will move back to Safari.

**Comment by kelimuttu** [comment]
We received feedback about this bug today via @FirefoxSupport on Twitter → https://twitter.com/donrmccurdy/status/1426655509289992193

**Comment by data-sync-user** [comment]
➤ Simion Basca commented:

Verified fix on v37.0 (5922).

**Comment by skabber** [comment]
Thank you @nishant2718 !
You have improved my life and that of others.

**Comment by naavis** [comment]
Not sure whether to open a new issue about this or not, but this issue is back. I am not yet sure how to reproduce it systematically, but quite often when I have my keyboard attached to my iPad and tap open a link on the Homepage, it opens it in another tab in the background.

**Comment by adamcstephens** [comment]
I too am experiencing this same problem again, and exactly as @naavis describes.

**Comment by gottoouu** [comment]
[
]()

**Comment by adamcstephens** [comment]
Clicking a link, opens the resulting link in a background tab:

https://user-images.githubusercontent.com/2071575/150681627-c7d23419-66bd-43fa-9452-6979fe026858.MOV

**Comment by naavis** [comment]
That is a good point. I notice that if I hold down Cmd while tapping a link, it opens it in a background tab, and many subsequent taps open a new background tab too eventhough I have already released the Cmd key. I can’t feel the Cmd key physically sticking down, but I guess there could be something to this theory.

**Comment by naavis** [comment]
Well, I did not find an app, but messed around with this: https://unixpapa.com/js/testkey.html

It always shows a keyup event when releasing the physical Cmd key, so I don’t think it is stuck. The only time I can see it not producing a keyup event is when I hold down Cmd long enough for Firefox to show the keyboard shortcut screen. But tapping on any links after that opens them correctly in the current tab.

So I can pretty reliably reproduce this now by doing one of the following:

1. Open the Homepage with Cmd + T (tapping the plus icon in the upper right corner is not enough) and tap on any link, or
2. Open the Homepage by tapping the plus icon in the upper right corner, then hold Cmd while tapping a link, then release Cmd and tap on a link again

I have the Apple Smart Keyboard, like I stated earlier.

**Comment by kelimuttu** [comment]
This should have been fixed in 37. Can you let us know which Firefox version you're using @naavis @adamcstephens ?

**Comment by naavis** [comment]
As I mentioned above:

> I am on Firefox Daylight 96.0 (7403).

**Comment by adamcstephens** [comment]
@kelimuttu I am also on Firefox Daylight 96.0 (7403).

One of the reasons I suggested it could be a stuck command key is that I can work around the bug by tapping the command key to “clear” it. Maybe this is happening at a low enough level that the test site can’t even see the stuck key?

**Comment by naavis** [comment]
I can confirm, tapping the Cmd key once stops the behaviour.

**Comment by lmarceau** [comment]
Hello! There was some recent changes on the keyboard shortcuts - I'll look into it. As you all said it indeed seems the command key is still registered as being pressed down. I think this happens when some combination of other shortcuts are used (and there's a tab change) - then the command key state isn't registered properly. This is annoying for sure, "fix" for now is to re-press the command key to reset it's state. This issue isn't exactly as the original one, but let's keep this here in all the same thread 👍 Thanks a lot for reporting this!

**Comment by danielmcormond** [comment]
Just another data point, I am experiencing the same behavior:
Version Daylight 96.0 (7403) and a logitech K380 bluetooth keyboard.

**Comment by jwilker** [comment]
One more data point. 
Firefox for ios (iPad Pro 10.5”) Version 97.

Clicking links opens new tabs at random, vs. following link. Same experience with cmd-r as others.

**Comment by walletje-w** [comment]
I’ve been experiencing this issue for quite some time now as well and was hoping that it would magically disappear with an update but so far no luck. 

Using iPad Pro 12.9” Third generation on iOS: 14.8, always connected to the Apple folio keyboard. 
Firefox: Daylight 98.2 (8475)

Can confirm that pressing the cmd buttons solves the issue temporarily.

**Comment by yoanarios** [comment]
After a lot of investigation this issue is being caused because pressesEnded at BVC is not being called intermittently for (cmd + T) and (cmd + W) so the KeyboardPressesHandler is returning Cmd being press when is not. 

I open 2 PRs to improve the behavior but I couldn’t pinpoint the root cause of the issue so it could be another cases triggering this. I will open another ticket to investigate but for 101 it's improved

**Comment by data-sync-user** [comment]
➤ Simion Basca commented:

Marking as verified fix on v101.0 (9723) and 9000 (9758). The issue is not reproducible when:

  * open new tab by using shortcut cmd + T.
  * closing Tab (cmd + W)
  * go forward (cmd + ])
  * reload a page (cmd+R" )
